pxtlib/analytics.ts

```diff
@@ -1,13 +1,60 @@
+/// <reference path="../localtypings/mscc" />
+
+namespace pxt {
+    // This function is defined in docfiles/pxtweb/cookieCompliance.ts
+    export declare let onAnalyticsLoaded: (cb: () => void) => void;
+}
+
 namespace pxt.analytics {
-    export function enable() {
-        let ai = (window as any).appInsights;
-        if (!ai) return;
+    // Completely arbitrarily chosen
+    const eventBufferSizeLimit = 20;
 
-        pxt.debug('enabling app insights')
+    interface TickData {
+        id: string;
+        data?: Map<string | number>;
+        opts?: TelemetryEventOptions;
+    }
+
+    interface ExceptionData {
+        err: any;
+        data: pxt.Map<string>;
+    }
+
+    interface ErrorData {
+        cat: string;
+        msg: string;
+        data?: pxt.Map<string>;
+    }
+
+    export function enable() {
+        pxt.debug('setting up app insights')
 
         const te = pxt.tickEvent;
-        pxt.tickEvent = function (id: string, data?: Map<string | number>): void {
-            if (te) te(id, data);
+        const teHandled = wrap((ev: TickData) => tickEventCore(ev.id, ev.data, ev.opts));
+        pxt.tickEvent = function (id: string, data?: Map<string | number>, opts?: TelemetryEventOptions): void {
+            teHandled({ id, data, opts });
+        };
+
+        const rexp = pxt.reportException;
+        const rexpHandled = wrap((ev: ExceptionData) => reportExceptionCore(ev.err, ev.data));
+        pxt.reportException = function (err: any, data: pxt.Map<string>): void {
```

Comment:
this function can be overriden, does it matter? it won't be using wrap most likely.

Reply:
I don't think it matters. Who overrides this BTW? The only place I could find in the codebase is the editor controller business and that shouldn't matter.

Reply:
If someone overrides it and they don't call the original then it's a moot point anyway.